lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs

```diff
@@ -254,26 +267,98 @@ instance ((PersistPublicKey (key 'AccountK)), Typeable chain)
     build (AddressPool acct (AddressPoolGap g) _) = mempty
         <> ccF <> " " <> acctF <> " (gap=" <> build g <> ")\n"
       where
-        ccF = build $ toText $ accountingStyle @chain
+        ccF = build $ toText $ role @chain
+        xpubF = hexF $ serializeXPub acct
+        acctF = prefixF 8 xpubF <> "..." <> suffixF 8 xpubF
+
+data VerificationKeyPool (key :: Depth -> * -> *) = VerificationKeyPool {
+      accountXPub
+        :: !(key 'AccountK XPub)
+    -- ^ Corresponding key for the pool (a pool is tied to only one account)
+    , keyNum
+        :: !AddressPoolGap
+    -- ^ The actual gap for the pool. This can't change for a given pool.
+    , indexedVerKeys
+        :: !(Map KeyHash (Index 'Soft 'ScriptK, AddressState))
+    } deriving (Generic)
+
+deriving instance (Show (key 'AccountK XPub))
+    => Show (VerificationKeyPool key)
+
+deriving instance (Eq (key 'AccountK XPub))
+    => Eq (VerificationKeyPool key)
+
+instance (NFData (key 'AccountK XPub))
+    => NFData (VerificationKeyPool key)
+
+instance ((PersistPublicKey (key 'AccountK)))
+    => Buildable (VerificationKeyPool key) where
+    build (VerificationKeyPool acct (AddressPoolGap g) _) = mempty
+       <> ccF <> " " <> acctF <> " (gap=" <> build g <> ")\n"
+      where
+        ccF = build $ toText MultisigScript
         xpubF = hexF $ serializeXPub acct
         acctF = prefixF 8 xpubF <> "..." <> suffixF 8 xpubF
 
--- | Bring a 'AccountingStyle' type back to the term-level. This requires a type
+instance Ord KeyHash where
+    compare (KeyHash kh1) (KeyHash kh2) =
+        compare kh1 kh2
+
+toVerKeyHash
+    :: WalletKey k
+    => k (depth:: Depth) XPub
+    -> KeyHash
+toVerKeyHash = KeyHash . blake2b224 . xpubPublicKey . getRawKey
+
+-- | Create a new VerificationKey pool.
+-- The extension to the pool is done by adding next adjacent indices,
+-- marking them as unused and their corresponding public keys.
+--
+mkVerificationKeyPool
+    :: (SoftDerivation k, WalletKey k)
+    => k 'AccountK XPub
+    -> AddressPoolGap
+    -> Map KeyHash (Index 'Soft 'ScriptK, AddressState)
```

Comment:
I thought to reuse it just to report whether the verification key was discovered or not. I thought it would be a good for a user to know if his keys were actually discovered and label that.